File: javascript/src/parser.ts
/**
 * @license
 * Copyright 2021 Google LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Base BudouX parser.
 */
export class Parser {
  /** BudouX model data */
  private readonly model: Map<string, Map<string, number>>;
  private readonly baseScore: number;

  /**
   * Constructs a BudouX parser.
   * @param model A model data.
   */
  constructor(model: {[key: string]: {[key: string]: number}}) {
    this.model = new Map(
      Object.entries(model).map(([k, v]: [string, {[key: string]: number}]) => [
        k,
        new Map(Object.entries(v)),
      ])
    );
    this.baseScore =
      -0.5 *
      [...this.model.values()]
        .flatMap((group: Map<string, number>) => [...group.values()])
        .reduce((prev: number, curr: number) => prev + curr, 0);
  }

  /**
   * Parses the input sentence and returns a list of semantic chunks.
   *
   * @param sentence An input sentence.
   * @return The retrieved chunks.
   */
  parse(sentence: string): string[] {
    if (sentence === '') return [];
    const boundaries = this.parseBoundaries(sentence);
    const result = [];
    let start = 0;
    for (const boundary of boundaries) {
      result.push(sentence.slice(start, boundary));
      start = boundary;
    }
    result.push(sentence.slice(start));
    return result;
  }

  /**
   * Parses the input sentence and returns a list of boundaries.
   *
   * @param sentence An input sentence.
   * @return The list of boundaries.
   */
  parseBoundaries(sentence: string): number[] {
    const result = [];
    const uw1 = this.model.get('UW1');
    const uw2 = this.model.get('UW2');
    const uw3 = this.model.get('UW3');
    const uw4 = this.model.get('UW4');
    const uw5 = this.model.get('UW5');
    const uw6 = this.model.get('UW6');
    const bw1 = this.model.get('BW1');
    const bw2 = this.model.get('BW2');
    const bw3 = this.model.get('BW3');
    const tw1 = this.model.get('TW1');
    const tw2 = this.model.get('TW2');
    const tw3 = this.model.get('TW3');
    const tw4 = this.model.get('TW4');

    for (let i = 1; i < sentence.length; i++) {
      let score = this.baseScore;
      // NOTE: Score values in models may be negative.
      score += uw1?.get(sentence.substring(i - 3, i - 2)) || 0;
      score += uw2?.get(sentence.substring(i - 2, i - 1)) || 0;
      score += uw3?.get(sentence.substring(i - 1, i)) || 0;
      score += uw4?.get(sentence.substring(i, i + 1)) || 0;
      score += uw5?.get(sentence.substring(i + 1, i + 2)) || 0;
      score += uw6?.get(sentence.substring(i + 2, i + 3)) || 0;
      score += bw1?.get(sentence.substring(i - 2, i)) || 0;
      score += bw2?.get(sentence.substring(i - 1, i + 1)) || 0;
      score += bw3?.get(sentence.substring(i, i + 2)) || 0;
      score += tw1?.get(sentence.substring(i - 3, i)) || 0;
      score += tw2?.get(sentence.substring(i - 2, i + 1)) || 0;
      score += tw3?.get(sentence.substring(i - 1, i + 2)) || 0;
      score += tw4?.get(sentence.substring(i, i + 3)) || 0;
      if (score > 0) result.push(i);
    }
    return result;
  }
}

File: golang/parser.go
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//	https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package budoux

import "github.com/google/budoux/golang/models"

type Parser struct {
	model models.Model
}

// Create new parser.
func New(model models.Model) *Parser {
	return &Parser{
		model: model,
	}
}

// NewDefaultJapaneseParser returns new Parser with default japanese model.
func NewDefaultJapaneseParser() *Parser {
	return New(models.GetDefaultJapaneseModel())
}

// NewDefaultThaiParser returns new Parser with default thai model.
func NewDefaultThaiParser() *Parser {
	return New(models.GetDefaultThaiModel())
}

// NewDefaultSimplifiedChineseParser returns new Parser with default simplified chinese model.
func NewDefaultSimplifiedChineseParser() *Parser {
	return New(models.GetDefaultSimplifiedChineseModel())
}

// NewDefaultTraditionalChineseParser returns new Parser with default traditional chinese model.
func NewDefaultTraditionalChineseParser() *Parser {
	return New(models.GetDefaultTraditionalChineseModel())
}

// getScore returns the score of a word in a group if exists.
// otherwise returns defaultScore.
func (p *Parser) getScore(group string, word string, defaultScore float64) float64 {
	if g, ok := p.model[group]; ok {
		if score, ok := g[word]; ok {
			//return float64(score)
			return score
		}
	}
	return defaultScore
}

// Parses a sentence into phrases.
func (p *Parser) Parse(sentence string) []string {
	if sentence == "" {
		return []string{}
	}
	runes := []rune(sentence)
	//baseScore := -float64(p.model.TotalScore()) * 0.5
	baseScore := -p.model.TotalScore() * 0.5
	chunks := []string{string(runes[0])}
	for i := 1; i < len(runes); i++ {
		score := baseScore
		if i > 2 {
			score += p.getScore("UW1", string(runes[i-3]), 0)
		}
		if i > 1 {
			score += p.getScore("UW2", string(runes[i-2]), 0)
		}
		score += p.getScore("UW3", string(runes[i-1]), 0)
		score += p.getScore("UW4", string(runes[i]), 0)
		if i+1 < len(runes) {
			score += p.getScore("UW5", string(runes[i+1]), 0)
		}
		if i+2 < len(runes) {
			score += p.getScore("UW6", string(runes[i+2]), 0)
		}

		if i > 1 {
			score += p.getScore("BW1", string(runes[i-2:i]), 0)
		}
		score += p.getScore("BW2", string(runes[i-1:i+1]), 0)
		if i+1 < len(runes) {
			score += p.getScore("BW3", string(runes[i:i+2]), 0)
		}

		if i > 2 {
			score += p.getScore("TW1", string(runes[i-3:i]), 0)
		}
		if i > 1 {
			score += p.getScore("TW2", string(runes[i-2:i+1]), 0)
		}
		if i+1 < len(runes) {
			score += p.getScore("TW3", string(runes[i-1:i+2]), 0)
		}
		if i+2 < len(runes) {
			score += p.getScore("TW4", string(runes[i:i+3]), 0)
		}

		if score > 0 {
			chunks = append(chunks, string(runes[i]))
		} else {
			chunks[len(chunks)-1] += string(runes[i])
		}
	}
	return chunks
}
